Lib/dataclasses.py

```diff
@@ -955,7 +956,11 @@ def _process_class(cls, init, repr, eq, order, unsafe_hash, frozen,
             has_dataclass_bases = True
             for f in base_fields.values():
                 fields[f.name] = f
-            if getattr(b, _PARAMS).frozen:
+            if all_frozen_bases is None:
+                all_frozen_bases = True
+            current_frozen = getattr(b, _PARAMS).frozen
+            all_frozen_bases = all_frozen_bases and current_frozen
+            if current_frozen:
                 any_frozen_base = True
```

Comment:
This would be better as
```suggestion
            any_frozen_base = any_frozen_base or current_frozen
```